app/store/migrations/079.ts

```
@@ -0,0 +1,209 @@
+import { cloneDeep, escapeRegExp } from 'lodash';
+
+const VERSION = 79;
+
+// Chains supported by Infura that are either built in or featured,
+// mapped to their corresponding failover URLs.
+// Copied from `PopularList` in app/util/networks/customNetworks.ts:
+// <https://github.com/MetaMask/metamask-mobile/blob/c29c22bf2ad62171c4cad3b4156500a1347aa7dc/app/util/networks/customNetworks.tsx#L12>
+export const INFURA_CHAINS_WITH_FAILOVERS: Map<
+  Hex,
+  { subdomain: string; getFailoverUrl: () => string | undefined }
+> = new Map([
+  [
+    '0x1',
+    {
+      subdomain: 'mainnet',
+      getFailoverUrl: () => process.env.QUICKNODE_MAINNET_URL,
+    },
+  ],
+  // linea mainnet
+  [
+    '0xe708',
+    {
+      subdomain: 'linea-mainnet',
+      getFailoverUrl: () => process.env.QUICKNODE_LINEA_MAINNET_URL,
+    },
+  ],
+  [
+    '0xa4b1',
+    {
+      subdomain: 'arbitrum',
+      getFailoverUrl: () => process.env.QUICKNODE_ARBITRUM_URL,
+    },
+  ],
+  [
+    '0xa86a',
+    {
+      subdomain: 'avalanche',
+      getFailoverUrl: () => process.env.QUICKNODE_AVALANCHE_URL,
+    },
+  ],
+  [
+    '0xa',
+    {
+      subdomain: 'optimism',
+      getFailoverUrl: () => process.env.QUICKNODE_OPTIMISM_URL,
+    },
+  ],
+  [
+    '0x89',
+    {
+      subdomain: 'polygon',
+      getFailoverUrl: () => process.env.QUICKNODE_POLYGON_URL,
+    },
+  ],
+  [
+    '0x2105',
+    {
+      subdomain: 'base',
+      getFailoverUrl: () => process.env.QUICKNODE_BASE_URL,
+    },
+  ],
+]);
+
+export default function migrate(state: unknown) {
+  const newState = cloneDeep(state);
+
+  try {
+    updateState(newState);
+    return newState;
+  } catch (error) {
+    captureException(
+      new Error(`FATAL ERROR: Migration ${VERSION}: ${getErrorMessage(error)}`),
+    );
+    return state;
+  }
+}
+
+function updateState(state: unknown) {
+  if (!process.env.MM_INFURA_PROJECT_ID) {
+    throw new Error('No MM_INFURA_PROJECT_ID set!');
+  }
+
+  if (!isObject(state)) {
```

Can we use [ensureValidState](https://github.com/MetaMask/metamask-mobile/blob/27d0ba8cafe40fa8e31479d7fb187774aa48cfee/app/store/migrations/util/index.tsx#L11) function ? It will cover the first 5 checks I believe, let me know your thoughts, if you think it's better to create different conditions between `hasProperty` and `isObject` and generate different errors, we could definitely update the generic function, so the next migrations be easier

---

Ah I see the issue sorry, you need it to throw instead of return state, ignore my comment, although It would be nice to keep following the existing pattern, or if we think on a better one, we could start improving it

---

Sure no problem. I've made use of `ensureValidState` here: [305d6cf](https://github.com/MetaMask/metamask-mobile/pull/15077/commits/305d6cfc7053684ee70f419d2a1ef30eec60c007)